[flang][Multi-Image] Moving Mutli-image lowering to PRIF into the MIF dialect

JDPailleux opened the pull request:
Support for multi-image features has begun to be integrated into LLVM. A new dialect which simplifies lowering to PRIF wil be proposed in this PR.
The initial definition of this dialect (MIF) is based only on operations already upstreamed in LLVM and the current lowering will be moved to this dialect.

Features like TEAMs and the use of coarrays will follow later in other PRs.
Any feedback regarding what is proposed is welcome.

JDPailleux committed:
[flang] Defining bases of the MIF Dialect

JDPailleux committed:
[flang][Multi-Image] Moving Mutli-image lowering to PRIF into the MIF Dialect

jeanPerier reviewed (commented):
Thanks @JDPailleux for adding a dialect for multi image Fortran!

It is nice to centralize the interface with PRIF in an MLIR pass and to have operations for MIF/Coarrays that will make experimenting easier.

I added comments inline, overall this looks great to me.

bonachea reviewed (commented):
@JDPailleux Thanks for all the great work on this new MIF dialect!

I don't follow all the LLVM details in this PR, but I've added some questions/comments on parts I think I understand.

kiranchandramohan commented:
> Thanks @JDPailleux for the progress with the Multi Image Dialect. Does this currently cover only the intrinsics?
> 
> If you could write a high level description of the MIF dialect as a PR to flang/docs or write a summary in discourse (https://discourse.llvm.org/c/subprojects/flang/33) that will make it easier for others to follow and help review. Alternatively you could consider presenting at the Flang community call next Wednesday.
> 
> Are there any general principles that you are following. Like all co-array or multi-image intrinsics will be represented as an operation. Are there any new FIR/MLIR types to represent coarrays/multi-images? Is the dialect only for ease of lowering or will it include some transformations as well? What kind of transformations do you plan to have in the dialect?

I see that you have some slides in https://flang-compiler.slack.com/archives/C06QTP07SCU/p1758900973869609. This addresses my points above. Thanks.

JDPailleux commented on flang/include/flang/Optimizer/Dialect/MIF/MIFOps.td:
```diff
@@ -0,0 +1,274 @@
+
+  let hasVerifier = 1;
+  let assemblyFormat = [{
+    ( `coarray` `(` $coarray^ `:` type($coarray) `)` )? 
```
For the pretty printer, I didn't know which approach to take because in FIR and CUF there are two different approaches used  (I use the CUF style), but in this case I'll go with the one you suggest except if there is other feedback before :)

JDPailleux committed:
Applying feedback

- Add decoration MemWrite and MemRead
- Remote useless --split-file flag
- Remote useless Start/End/Main in the conversion tests
- Replace PRIF macro type by static functions
- Remove unecessary things

JDPailleux committed:
Update flang/include/flang/Optimizer/Dialect/MIF/MIFOps.td

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

JDPailleux committed:
Update flang/lib/Optimizer/Dialect/MIF/MIFOps.cpp

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

JDPailleux committed:
Update flang/lib/Optimizer/Dialect/MIF/MIFOps.cpp

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

clementval commented on flang/include/flang/Optimizer/Dialect/MIF/MIFDialect.h:
```diff
@@ -0,0 +1,31 @@
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Interfaces/VectorInterfaces.h"
+
+// #include "flang/Optimizer/Dialect/FIRAttr.h"
+// #include "flang/Optimizer/Dialect/FIRType.h"
```
Remove?

clementval commented on flang/include/flang/Optimizer/Dialect/MIF/MIFDialect.h:
```diff
@@ -0,0 +1,31 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_OPTIMIZER_DIALECT_MIF_MIF_H
+#define FORTRAN_OPTIMIZER_DIALECT_MIF_MIF_H
```
```suggestion
#ifndef FORTRAN_OPTIMIZER_DIALECT_MIF_MIFDIALECT_H
#define FORTRAN_OPTIMIZER_DIALECT_MIF_MIFDIALECT_H
```

clementval commented on flang/include/flang/Optimizer/Dialect/MIF/MIFOps.td:
```diff
@@ -0,0 +1,267 @@
+//===-- MIFOps.td - MIF operation definitions ------*- tablegen -*-===//
```
Line should be 80 chars. You are missing couple of `-`

bonachea reviewed (commented):
@JDPailleux Thanks for all the improvements!

I'm very excited to see this effort progressing.

JDPailleux committed:
Update flang/include/flang/Optimizer/Dialect/MIF/MIFDialect.h

Co-authored-by: Valentin Clement (バレンタイン クレメン) <clementval@gmail.com>

JDPailleux committed:
Update flang/include/flang/Optimizer/Dialect/MIF/MIFDialect.h

Co-authored-by: Valentin Clement (バレンタイン クレメン) <clementval@gmail.com>

JDPailleux committed:
Update assemblyFormat

JDPailleux committed:
Remove useless comments header

jeanPerier reviewed (approved):
Please take care of the missing side effect, LGTM otherwise. Thanks for adding this dialect!

bonachea reviewed (commented):
All the recent improvements look great!

I added a few more observations to consider as possible improvements.

jeanPerier commented on flang/include/flang/Optimizer/Dialect/MIF/MIFOps.td:
```diff
@@ -0,0 +1,263 @@
+//===----------------------------------------------------------------------===//
+
+def mif_SyncAllOp : mif_Op<"sync_all", [AttrSizedOperandSegments,
+                                        MemoryEffects<[MemWrite]>]> {
```
It is a good point that a sync operation will not only write but also read something from the runtime state.
`FullEffect` seems to imply all the memory resource is being read/written. It is not the case here, and if anything, I would imagine this is an indication that allows doing more optimizations (like deleting the first write in write after write). It is not used on any of the FIR operations.

The safest approach is to not implement the side effect interface at all, this will prevent any dialect independent optimizations involving memory. My understanding is that this is semantically equivalent to specifying `MemoryEffects<[MemRead,MemWrite,MemAlloc,MemFree]>`, with the exception that the latter imply that the operation has no memory conflict with an operation that would have MemoryEffects on something else than the DefaultRessource.

For now I would advise starting by not implementing the side effect interface at all without some deeper design and testing phase.

I checked the [MPI dialect](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/MPI/IR/MPIOps.td) and it does not implement side effect interfaces outside of few invariant operations that they consider Pure.

Later, there is maybe a case from ensuring some optimizations like load to store forwarding of local scalars that are not involved in any multi image statements is still possible (my understanding is that an operation that either does not implement the side effect interface or has generic write effect will prevent such optimization if it is between the load and store, even if the operation does not take the address, or an alias of the address as an argument). We should likely have a discussion with the MPI dialect contributors when trying to specify side effects in a finer way.

bonachea replied:
This thread is now resolved.

JDPailleux committed:
Adding MemRead memory effect + add verifier to CoBroadcastOp + code factorization

JDPailleux committed:
Removing memory effects on synchonrization ops

bonachea reviewed (commented):
Made one minor cosmetic request, otherwise LGTM!

ktras reviewed (approved):
I have verified through testing with Caffeine that this PR doesn't contain any regressions for the features switched from the experimental support to support using the new MIF dialect.

JDPailleux committed:
Update flang/include/flang/Optimizer/Dialect/MIF/MIFOps.td

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

bonachea reviewed (commented):
LGTM! Thanks for all the work and improvements!
